src/transformers/models/distilbert/modeling_distilbert.py

```diff
@@ -174,15 +195,54 @@ def unshape(x):
+        # k = shape(self.k_lin(key))  # (bs, n_heads, k_length, dim_per_head)
+        # v = shape(self.v_lin(value))  # (bs, n_heads, k_length, dim_per_head)
+
+        # If this is instantiated as a cross-attention module, the keys
+        # and values come from an encoder; the attention mask needs to be
+        # such that the encoder's padding tokens are not attended to.
+        is_cross_attention = encoder_hidden_states is not None
+
+        if is_cross_attention and past_key_values is not None:
+            # reuse k,v, cross_attentions
+            k = past_key_values[0]
+            v = past_key_values[1]
+            attention_mask = encoder_attention_mask
+        elif is_cross_attention:
+            k = shape(self.k_lin(key))  # (bs, n_heads, k_length, dim_per_head)
+            v = shape(self.v_lin(value))  # (bs, n_heads, k_length, dim_per_head)
+            attention_mask = encoder_attention_mask
+        elif past_key_values is not None:
+            k = shape(self.k_lin(key))  # (bs, n_heads, k_length, dim_per_head)
+            v = shape(self.v_lin(value))  # (bs, n_heads, k_length, dim_per_head)
+            k = torch.cat([past_key_values[0], k], dim=2)
+            v = torch.cat([past_key_values[1], v], dim=2)
+        else:
+            k = shape(self.k_lin(key))  # (bs, n_heads, k_length, dim_per_head)
+            v = shape(self.v_lin(value))  # (bs, n_heads, k_length, dim_per_head)
+
+        if self.is_decoder:
+            # if cross_attention save Tuple(torch.Tensor, torch.Tensor) of all cross attention key/value_states.
+            # Further calls to cross_attention layer can then reuse all cross-attention
+            # key/value_states (first "if" case)
+            # if uni-directional self-attention (decoder) save Tuple(torch.Tensor, torch.Tensor) of
+            # all previous decoder key/value_states. Further calls to uni-directional self-attention
+            # can concat previous decoder key/value_states to current projected key/value_states (third "elif" case)
+            # if encoder bi-directional self-attention `past_key_value` is always `None`
+            past_key_values = (k, v)
+
         q = shape(self.q_lin(query))  # (bs, n_heads, q_length, dim_per_head)
-        k = shape(self.k_lin(key))  # (bs, n_heads, k_length, dim_per_head)
-        v = shape(self.v_lin(value))  # (bs, n_heads, k_length, dim_per_head)
 
         q = q / math.sqrt(dim_per_head)  # (bs, n_heads, q_length, dim_per_head)
         scores = torch.matmul(q, k.transpose(2, 3))  # (bs, n_heads, q_length, k_length)
         mask = (mask == 0).view(mask_reshp).expand_as(scores)  # (bs, n_heads, q_length, k_length)
         scores.masked_fill_(mask, -float("inf"))  # (bs, n_heads, q_length, k_length)
 
+        if attention_mask is not None:
+            # Apply the attention mask is (precomputed for all layers in DistilBertModel forward() function)
+            scores = scores + attention_mask
```

Comment:
as commented above `mask` and `attention_mask` are similar, so this can be removed.